doc/compiler/aot/InlinedMethods.md

```diff
@@ -0,0 +1,259 @@
+[Intro to AOT blog series](https://blog.openj9.org/2018/10/10/intro-to-ahead-of-time-compilation/),
+which this doc assumes the reader has already read.
+
+# Introduction
+
+There are four general kinds of inlined method relocations:
+
+* Inlined Methods
+* Inlined Methods with NOP Guards
+* Profiled Inlined Methods
+* Profiled Inlined Methods with Guards
+
+All relocations are used to validate the inlined method. However, the
+relocations for methods with guards are used to relocate the associated
+inlined table entry in the metadata as well as the guards for the 
+inlined sites, whereas the rest are only used to relocate the associated
+inlined table entry. Profiled Inlined Methods are methods that were 
+inlined that were not named by the caller. For example,
+consider the following scenario:
+
+```
+class base {
+	void foo();
+}
+class child_one extends base {
+	void foo();
+}
+child class_two extends base {
+	void foo();
+}
+...
+	void some_method(base b) {
+		b.foo();
+	}
+```
+Because `b` is of type `base`, the `foo` that gets executed could either
+be the implementation in `base` or `child_one` or `child_two`. If the
+inliner decided to pick `base.foo()` because that is what was named in
+`some_method`, then the relocation would be the Inlined Method variant.
+However, if the inliner decided to inline `child_two.foo()` based on 
+profiling information, the relocation would be the Profiled Inlined
+Method variant.
+
+# Generating Guards & External Relocations
+
+During an AOT compilation, i.e. when the code is first generated, external
+relocation are generated at various points. Once all of these external 
+relocations are added to a list, the inlined method external relocations
+are then added in reverse order; because entries are added to the
+list at the front, an external relocation associated with the last inlined
+site index is added first, followed by the second last, continuing to the
+first. This ensures that the first inlined method external relocation in
+the list is associated with the first inlined site.
+
+The reason for ensuring this is because validating/materialize values for
+subsequent inlined methods may depend on the previous entry. For example,
+if `foo() calls bar() calls baz()`, and `bar` and `baz` are inlined (first
+`bar` would be inlined, then `baz`), `baz` is named in `bar`'s constant
+pool, and `bar` is named in `foo`'s constant pool. Therefore, in order to
+materialize the `J9Method` of `baz`, one first needs the `J9Method` of `bar`,
+which first requires the `J9Method` of `foo`.
+
+Generating external relocations for inlined methods requires the 
+existence of Guards. Guards are used to gate execution of regions of 
+code. When a method is inlined, a guard is generated for that inlined 
+method. It is possible for a guard to be removed for various reasons. 
+Currently, an inlined method cannot be validated/relocated without an 
+associated guard<sup>1</sup>. As such, if a guard is removed, it is 
```

Comment:
I think we should put the footnote at the end of the section so the reader doesn't have to navigate to the end of the document for it. Think of where footnotes are placed in a book. Typically they are at the bottom of the current page. We don't have "pages" here per say, so sections is the next best thing.

Also I think you should be able to do:

<sup>1</sup> This is a problem that will be fixed soon<sup>TM</sup> as features like OSR, NextGenHCR, and FSD cannot otherwise be supported for AOT. 

Moreover though I would like to see this footnote backed by an issue, otherwise it will never get fixed. The issue should also backpoint to this article to remove the footnote once it is actually fixed so we don't forget to update the documentation. (or is fixing this a true pipedream?)

Reply:
> Moreover though I would like to see this footnote backed by an issue, otherwise it will never get fixed. The issue should also backpoint to this article to remove the footnote once it is actually fixed so we don't forget to update the documentation. (or is fixing this a true pipedream?)

No there's issues for it; I'm also actively working on FSD for AOT so I definitely already have the fix written up; PRs will be open in the near future at which point I can update this doc to reflect that fact.